In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [7]:
file_path = '~/Desktop/Products.csv'

product_df = pd.read_csv(file_path)

# DATA WRANGLING

First we'll start by importing the 'Product' data that contains information pertaining specifically to the products our retailer sells. We will investigate the following before:

- Shape of the dataset 
- Missing Values 
- Number of unique products, subcategories, categories 
- Range of unit cost and unit price

Other information about things like the brand and types of products will be uncovered at a later time when we have a fully working dataset 

# Product DataFrame

In [8]:
product_df.head()

product_key                         product_name    brand   color  \
0            1  Contoso 512MB MP3 Player E51 Silver  Contoso  Silver   
1            2    Contoso 512MB MP3 Player E51 Blue  Contoso    Blue   
2            3     Contoso 1G MP3 Player E100 White  Contoso   White   
3            4    Contoso 2G MP3 Player E200 Silver  Contoso  Silver   
4            5       Contoso 2G MP3 Player E200 Red  Contoso     Red   

  unit_cost_usd unit_price_usd  subcategory_key subcategory  category_key  \
0        $6.62         $12.99               101     MP4&MP3             1   
1        $6.62         $12.99               101     MP4&MP3             1   
2        $7.40         $14.52               101     MP4&MP3             1   
3       $11.00         $21.57               101     MP4&MP3             1   
4       $11.00         $21.57               101     MP4&MP3             1   

  category  
0    Audio  
1    Audio  
2    Audio  
3    Audio  
4    Audio

In [9]:
product_df.shape

(2517, 10)

In [11]:
product_df['product_key'].nunique()

2517

This is the sign of a large retailr as they sell 2,517 unique electronic products.

In [12]:
product_df.subcategory.nunique()

32

In [13]:
product_df.subcategory.unique()

array(['MP4&MP3', 'Recording Pen', 'Bluetooth Headphones', 'Televisions',
       'VCD & DVD', 'Home Theater System', 'Car Video', 'Laptops',
       'Desktops', 'Monitors', 'Projectors & Screens',
       'Printers, Scanners & Fax', 'Computers Accessories',
       'Digital Cameras', 'Digital SLR Cameras', 'Camcorders',
       'Cameras & Camcorders Accessories', 'Home & Office Phones',
       'Touch Screen Phones', 'Smart phones & PDAs', 'Movie DVD',
       'Boxed Games', 'Download Games', 'Washers & Dryers',
       'Refrigerators', 'Microwaves', 'Water Heaters', 'Coffee Machines',
       'Lamps', 'Air Conditioners', 'Fans', 'Cell phones Accessories'],
      dtype=object)

In [14]:
product_df.isna().sum()

product_key        0
product_name       0
brand              0
color              0
unit_cost_usd      0
unit_price_usd     0
subcategory_key    0
subcategory        0
category_key       0
category           0
dtype: int64

There are no missing values in the product_df, which is a great sign.

In [15]:
product_df.describe()

product_key  subcategory_key  category_key
count  2517.000000      2517.000000   2517.000000
mean   1259.000000       491.810091      4.878824
std     726.739637       229.887134      2.299170
min       1.000000       101.000000      1.000000
25%     630.000000       305.000000      3.000000
50%    1259.000000       406.000000      4.000000
75%    1888.000000       801.000000      8.000000
max    2517.000000       808.000000      8.000000

In [16]:
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2517 entries, 0 to 2516
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   product_key      2517 non-null   int64 
 1   product_name     2517 non-null   object
 2   brand            2517 non-null   object
 3   color            2517 non-null   object
 4   unit_cost_usd    2517 non-null   object
 5   unit_price_usd   2517 non-null   object
 6   subcategory_key  2517 non-null   int64 
 7   subcategory      2517 non-null   object
 8   category_key     2517 non-null   int64 
 9   category         2517 non-null   object
dtypes: int64(3), object(7)
memory usage: 196.8+ KB


It looks like unit cost and unit price are object types and should be numeric. This is likely because the data contains a '$' before the amount. We'll clean this now:

In [17]:
product_df[['unit_cost_usd', 'unit_price_usd']].head()

unit_cost_usd unit_price_usd
0        $6.62         $12.99 
1        $6.62         $12.99 
2        $7.40         $14.52 
3       $11.00         $21.57 
4       $11.00         $21.57

To resolve this, we need to:

1. Remove the dollar sign.
2. Remove any commas.
3. Strip any leading or trailing whitespace.
4. Finally, convert to float.

In [18]:
product_df['unit_cost_usd'] = (
    product_df['unit_cost_usd']
    .replace({'\$': '', ',':'', ' ':''},regex=True) #This ensures that all specified characters are removed in one go.
    .astype(float)
)

In [19]:
product_df['unit_cost_usd'].head()

0     6.62
1     6.62
2     7.40
3    11.00
4    11.00
Name: unit_cost_usd, dtype: float64

In [20]:
product_df['unit_price_usd'] = (
    product_df['unit_price_usd']
    .replace({'\$':'', ',':'', ' ':''}, regex=True)
    .astype(float)
)

In [21]:
product_df['unit_price_usd'].head()

0    12.99
1    12.99
2    14.52
3    21.57
4    21.57
Name: unit_price_usd, dtype: float64

In [22]:
product_df.describe()

product_key  unit_cost_usd  unit_price_usd  subcategory_key  \
count  2517.000000    2517.000000     2517.000000      2517.000000   
mean   1259.000000     147.655562      356.830131       491.810091   
std     726.739637     184.373683      494.054962       229.887134   
min       1.000000       0.480000        0.950000       101.000000   
25%     630.000000      32.250000       69.000000       305.000000   
50%    1259.000000      86.670000      199.990000       406.000000   
75%    1888.000000     183.950000      410.000000       801.000000   
max    2517.000000    1060.220000     3199.990000       808.000000   

       category_key  
count   2517.000000  
mean       4.878824  
std        2.299170  
min        1.000000  
25%        3.000000  
50%        4.000000  
75%        8.000000  
max        8.000000

We can see now that the unit cost and unit price columns have been converted into numeric format. The statistical summary of these two columns suggests a wide range of values, which typically indicates that there is a lot of variance. We can see that this company sells products as cheap as 95cents USD to over 3,000.

Once more we will check if there are missing values to ensure nothing went wrong during our conversion:

In [23]:
product_df.isna().sum()

product_key        0
product_name       0
brand              0
color              0
unit_cost_usd      0
unit_price_usd     0
subcategory_key    0
subcategory        0
category_key       0
category           0
dtype: int64

To recap the Product Dataframe:

- **product_key**         :Primary key to identify products (int)
- **product_name**        :Product Name (Object)
- **brand**               :Product Brand (Object)
- **color**               :Product Color (Object)
- **unit_cost_usd**       :Cost to produce product in USD (float)
- **unit_price_usd**      :Product list price in USD (float)
- **subcategory_key**     :Key to identify product subcategories (int)
- **ubcategory**         :Subcategory Name (Object)
- **categorykey**         :Key to identify product categories (int)
- **category**            :Product Category Name (Object)


# Sales DataFrame

Now we want to import the Sales DataFrame and inspect it as we did before. The Sales data containes the actual transaction history of our retailer. Here we'll find out the quantities that were actually sold, tied to unique customers and other information such as the store locations, order dates etc. 

In [36]:
file_path = '~/Desktop/Sales.csv'

sales_df = pd.read_csv(file_path)

In [37]:
sales_df.head()

order_number  line_item order_date delivery_date  customer_key  store_key  \
0        366000          1   1/1/2016           NaN        265598         10   
1        366001          1   1/1/2016     1/13/2016       1269051          0   
2        366001          2   1/1/2016     1/13/2016       1269051          0   
3        366002          1   1/1/2016     1/12/2016        266019          0   
4        366002          2   1/1/2016     1/12/2016        266019          0   

   product_key  quantity currency_code  
0         1304         1           CAD  
1         1048         2           USD  
2         2007         1           USD  
3         1106         7           CAD  
4          373         1           CAD

One thing we can flag right away is that our retailer operates across different countries, as we can see both USD and CAD are listed for currencies. We will inspect this further and we might have to create a column to convert currencies isn't to one baseline for easier comparison, and another column to ensure we preserve the location of stores.

But first, like we did before, here is our preliminary checklist to explore the dataframe:

- Shape of the dataset
- Missing Values
- Summary Statistics

In [38]:
sales_df.shape

(62884, 9)

In [39]:
sales_df.isna().sum()

order_number         0
line_item            0
order_date           0
delivery_date    49719
customer_key         0
store_key            0
product_key          0
quantity             0
currency_code        0
dtype: int64

Here we can see that delivery date is missing a staggering 79% of the data. If this was a real client we were working for we would definitely flag this and raise the issue to the client to inspect why there are missing delivery date values. This may be something the business would want to correct for. 

However, for our analysis, given how much data is missing in this column, we will delete it entirely as it will not help us with our analysis. 

In [40]:
sales_df.drop(columns='delivery_date', inplace=True)

In [41]:
sales_df

order_number  line_item order_date  customer_key  store_key  \
0            366000          1   1/1/2016        265598         10   
1            366001          1   1/1/2016       1269051          0   
2            366001          2   1/1/2016       1269051          0   
3            366002          1   1/1/2016        266019          0   
4            366002          2   1/1/2016        266019          0   
...             ...        ...        ...           ...        ...   
62879       2243030          1  2/20/2021       1216913         43   
62880       2243031          1  2/20/2021        511229          0   
62881       2243032          1  2/20/2021        331277          0   
62882       2243032          2  2/20/2021        331277          0   
62883       2243032          3  2/20/2021        331277          0   

       product_key  quantity currency_code  
0             1304         1           CAD  
1             1048         2           USD  
2             2007         1           USD  
3             1106         7           CAD  
4              373         1           CAD  
...            ...       ...           ...  
62879          632         3           USD  
62880           98         4           EUR  
62881         1613         2           CAD  
62882         1717         2           CAD  
62883          464         7           CAD  

[62884 rows x 8 columns]

In [42]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62884 entries, 0 to 62883
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_number   62884 non-null  int64 
 1   line_item      62884 non-null  int64 
 2   order_date     62884 non-null  object
 3   customer_key   62884 non-null  int64 
 4   store_key      62884 non-null  int64 
 5   product_key    62884 non-null  int64 
 6   quantity       62884 non-null  int64 
 7   currency_code  62884 non-null  object
dtypes: int64(6), object(2)
memory usage: 3.8+ MB


In [43]:
sales_df.describe()

order_number     line_item  customer_key     store_key   product_key  \
count  6.288400e+04  62884.000000  6.288400e+04  62884.000000  62884.000000   
mean   1.430905e+06      2.164207  1.180797e+06     31.802144   1125.859344   
std    4.532963e+05      1.365170  5.859634e+05     22.978188    709.244010   
min    3.660000e+05      1.000000  3.010000e+02      0.000000      1.000000   
25%    1.121017e+06      1.000000  6.808580e+05      8.000000    437.000000   
50%    1.498016e+06      2.000000  1.261200e+06     37.000000   1358.000000   
75%    1.788010e+06      3.000000  1.686496e+06     53.000000   1650.000000   
max    2.243032e+06      7.000000  2.099937e+06     66.000000   2517.000000   

           quantity  
count  62884.000000  
mean       3.144790  
std        2.256371  
min        1.000000  
25%        1.000000  
50%        2.000000  
75%        4.000000  
max       10.000000

The average quantity of units sold is 3 per order and our maxium sold is 10. Of course the minimum is 1 to indicate that a sales has been made. 

We can take a quick sneak peak at the unique currency types and get a sense for which currency has the most amount of quantity sold (we will explore relationships further later on):

In [44]:
sales_df.store_key.nunique()

58

In [45]:
sales_df.customer_key.nunique()

11887

In [46]:
sales_df.product_key.nunique()

2492

Our company is selling across 58 different stores, to 11,887 unique customers, in 5 different currencies and 2,492 unique products have been sold. 

This is a good time to note that in our previous 'Products' Dataframe, we uncovered that we have 2517 unique products; **which means 25 products are not generating sales.** 

We will make a note of that and investigate these products later .

# Merging DataFrames

So far we have taken a look at two big and important DataFrames 'Product' & 'Sales'. 

Now we'll merge these two dataframes so that we can begin to create our analysis. This will include inspecting features with EDA and creating some new features as well to further our analysis. 

In [47]:
merge_df = pd.merge(sales_df, product_df, on='product_key', how='inner')

merge_df.head()

order_number  line_item  order_date  customer_key  store_key  product_key  \
0        366000          1    1/1/2016        265598         10         1304   
1        378002          2   1/13/2016       1599716         45         1304   
2        868008          2   5/17/2017       1540067         51         1304   
3       1078002          1  12/13/2017        631631         18         1304   
4       1371000          1   10/2/2018       1257599         48         1304   

   quantity currency_code                     product_name    brand  color  \
0         1           CAD  Contoso Lens Adapter M450 White  Contoso  White   
1         1           USD  Contoso Lens Adapter M450 White  Contoso  White   
2         1           USD  Contoso Lens Adapter M450 White  Contoso  White   
3         1           EUR  Contoso Lens Adapter M450 White  Contoso  White   
4         8           USD  Contoso Lens Adapter M450 White  Contoso  White   

   unit_cost_usd  unit_price_usd  subcategory_key  \
0          31.27            68.0              406   
1          31.27            68.0              406   
2          31.27            68.0              406   
3          31.27            68.0              406   
4          31.27            68.0              406   

                        subcategory  category_key                category  
0  Cameras & Camcorders Accessories             4  Cameras and camcorders  
1  Cameras & Camcorders Accessories             4  Cameras and camcorders  
2  Cameras & Camcorders Accessories             4  Cameras and camcorders  
3  Cameras & Camcorders Accessories             4  Cameras and camcorders  
4  Cameras & Camcorders Accessories             4  Cameras and camcorders

In [48]:
merge_df.loc[merge_df.customer_key == 265598.0]

order_number  line_item  order_date  customer_key  store_key  \
0            366000          1    1/1/2016        265598         10   
15175       1749017          1  10/15/2019        265598          0   
23561        891000          2    6/9/2017        265598          9   
29251        891000          1    6/9/2017        265598          9   
32924        891000          4    6/9/2017        265598          9   
41012       1749017          3  10/15/2019        265598          0   
50666       1749017          2  10/15/2019        265598          0   
53702        891000          3    6/9/2017        265598          9   

       product_key  quantity currency_code  \
0             1304         1           CAD   
15175         1619         1           CAD   
23561          174         1           CAD   
29251          385         2           CAD   
32924           87         1           CAD   
41012           52         1           CAD   
50666         1764         2           CAD   
53702          685         1           CAD   

                                            product_name  \
0                        Contoso Lens Adapter M450 White   
15175                       Contoso DVD Player M130 Grey   
23561                               SV 22xDVD M600 Black   
29251                Adventure Works Laptop8.9 E0890 Red   
32924  NT Wireless Bluetooth Stereo Headphones M402 P...   
41012                WWI 2GB Pulse Smart pen M100 Silver   
50666       MGS Age of Empires II: The Age of Kings E126   
53702  Proseware Desk Jet All-in-One Printer, Scanner...   

                      brand   color  unit_cost_usd  unit_price_usd  \
0                   Contoso   White          31.27           68.00   
15175               Contoso    Grey          27.59           59.99   
23561      Southridge Video   Black          43.04          129.90   
29251       Adventure Works     Red         166.20          326.00   
32924     Northwind Traders  Purple          45.98           99.99   
41012  Wide World Importers  Silver          91.95          199.95   
50666         Tailspin Toys    Blue          23.86           46.80   
53702             Proseware    Grey          72.66          158.00   

       subcategory_key                       subcategory  category_key  \
0                  406  Cameras & Camcorders Accessories             4   
15175              602                         Movie DVD             6   
23561              202                         VCD & DVD             2   
29251              301                           Laptops             3   
32924              106              Bluetooth Headphones             1   
41012              104                     Recording Pen             1   
50666              702                    Download Games             7   
53702              306          Printers, Scanners & Fax             3   

                            category  
0             Cameras and camcorders  
15175  Music, Movies and Audio Books  
23561                   TV and Video  
29251                      Computers  
32924                          Audio  
41012                          Audio  
50666                 Games and Toys  
53702                      Computers

Here we'll make a note that the 'order_number' is not each unique purchase of a product. As we see, this particular customer had bought 4 different products on the same purchase, and they are seperated by 'line_item'. 

We may need to toggle with this later on in the analysis, as one order_number could also contain more than 1 quantity of that product. 

In [50]:
print(product_df.shape)
print(sales_df.shape)
print(merge_df.shape)

(2517, 10)
(62884, 8)
(62884, 17)


In [51]:
merge_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 62884 entries, 0 to 62883
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   order_number     62884 non-null  int64  
 1   line_item        62884 non-null  int64  
 2   order_date       62884 non-null  object 
 3   customer_key     62884 non-null  int64  
 4   store_key        62884 non-null  int64  
 5   product_key      62884 non-null  int64  
 6   quantity         62884 non-null  int64  
 7   currency_code    62884 non-null  object 
 8   product_name     62884 non-null  object 
 9   brand            62884 non-null  object 
 10  color            62884 non-null  object 
 11  unit_cost_usd    62884 non-null  float64
 12  unit_price_usd   62884 non-null  float64
 13  subcategory_key  62884 non-null  int64  
 14  subcategory      62884 non-null  object 
 15  category_key     62884 non-null  int64  
 16  category         62884 non-null  object 
dtypes: float64(2

The inner join ensured that we got records that have maching values in both tables. 

# Stores DataFrame

In [52]:
file_path = '~/Desktop/Stores.csv'

stores_df = pd.read_csv(file_path)

We'll also import and merge the stores dataframe so that we can include information on country, state, size of the store in square metres and how long the it has been opened to get a further sense of how the retailer is performing across different dimensions.

Let's quickly explore this DataFrame as well before we merge it into our final working DF:

- Shape of the dataset
- Missing Values
- Summary Statistics


In [53]:
stores_df.head()

store_key    country                         state  square_meters  \
0          1  Australia  Australian Capital Territory          595.0   
1          2  Australia            Northern Territory          665.0   
2          3  Australia               South Australia         2000.0   
3          4  Australia                      Tasmania         2000.0   
4          5  Australia                      Victoria         2000.0   

   open_date  
0   1/1/2008  
1  1/12/2008  
2   1/7/2012  
3   1/1/2010  
4  12/9/2015

In [54]:
rename_stores_columns = ['store_key', 'country', 'state', 'square_meters', 'open_date']

stores_df.columns = rename_stores_columns
stores_df.head()


store_key    country                         state  square_meters  \
0          1  Australia  Australian Capital Territory          595.0   
1          2  Australia            Northern Territory          665.0   
2          3  Australia               South Australia         2000.0   
3          4  Australia                      Tasmania         2000.0   
4          5  Australia                      Victoria         2000.0   

   open_date  
0   1/1/2008  
1  1/12/2008  
2   1/7/2012  
3   1/1/2010  
4  12/9/2015

In [55]:
stores_df.shape

(67, 5)

In [56]:
stores_df.isna().sum()

store_key        0
country          0
state            0
square_meters    1
open_date        0
dtype: int64

In [57]:
missing_rows = stores_df.loc[stores_df.isna().any(axis=1)]
missing_rows

store_key country   state  square_meters open_date
66          0  Online  Online            NaN  1/1/2010

It's a good thing that we investigated the missing value. In this case, square_meters are NaN because this is our online distribution channel and not a real store. We will leave this value as NaN for now, as later this column will be encoded in order for our models to work.

In [58]:
stores_df.country.unique()

array(['Australia', 'Canada', 'France', 'Germany', 'Italy', 'Netherlands',
       'United Kingdom', 'United States', 'Online'], dtype=object)

In [59]:
stores_df.state.unique()

array(['Australian Capital Territory', 'Northern Territory',
       'South Australia', 'Tasmania', 'Victoria', 'Western Australia',
       'New Brunswick', 'Newfoundland and Labrador',
       'Northwest Territories', 'Nunavut', 'Yukon', 'Basse-Normandie',
       'Corse', 'Franche-Comté', 'La Réunion', 'Limousin', 'Martinique',
       'Mayotte', 'Berlin', 'Brandenburg', 'Freie Hansestadt Bremen',
       'Freistaat Thüringen', 'Hamburg', 'Hessen',
       'Mecklenburg-Vorpommern', 'Saarland', 'Sachsen-Anhalt',
       'Caltanissetta', 'Enna', 'Pesaro', 'Drenthe', 'Flevoland',
       'Friesland', 'Groningen', 'Zeeland', 'Armagh', 'Ayrshire',
       'Belfast', 'Blaenau Gwent', 'Dungannon and South Tyrone',
       'Fermanagh', 'North Down', 'Alaska', 'Arkansas', 'Connecticut',
       'Delaware', 'Hawaii', 'Idaho', 'Iowa', 'Kansas', 'Maine',
       'Mississippi', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Mexico', 'North Dakota', 'Oregon', 'Rhode Island',
       'South Carol

In [60]:
stores_df.state.nunique()

67

In [61]:
stores_df.describe()

store_key  square_meters
count  67.000000      66.000000
mean   33.000000    1402.196970
std    19.485037     576.404058
min     0.000000     245.000000
25%    16.500000    1108.750000
50%    33.000000    1347.500000
75%    49.500000    2000.000000
max    66.000000    2105.000000

# Second Merge

In [62]:
df = pd.merge(merge_df, stores_df, on='store_key', how='inner')

In [63]:
df.head(10)

order_number  line_item  order_date  customer_key  store_key  product_key  \
0        366000          1    1/1/2016        265598         10         1304   
1       1896020          1   3/10/2020        302645         10          373   
2       1298021          1   7/21/2018        363584         10          163   
3       1506050          4   2/14/2019        391900         10          163   
4       1452075          3  12/22/2018        211746         10          421   
5       1658017          4   7/16/2019        285552         10          421   
6       1878036          2   2/21/2020        213985         10          421   
7        675012          2   11/5/2016        205476         10         1617   
8       1001002          2   9/27/2017        351430         10         1617   
9       1625012          4   6/13/2019        398671         10         2510   

   quantity currency_code                                       product_name  \
0         1           CAD                    Contoso Lens Adapter M450 White   
1         1           CAD              Adventure Works Laptop8.9 E0890 White   
2         8           CAD           Adventure Works 52" LCD HDTV X790W White   
3         4           CAD           Adventure Works 52" LCD HDTV X790W White   
4         2           CAD         Adventure Works Desktop PC3.0 MS300 Silver   
5         1           CAD         Adventure Works Desktop PC3.0 MS300 Silver   
6         1           CAD         Adventure Works Desktop PC3.0 MS300 Silver   
7         2           CAD                     Contoso DVD Player M110 Silver   
8         6           CAD                     Contoso DVD Player M110 Silver   
9         5           CAD  Contoso Original K1m Li-Ion Standard Battery E...   

             brand  ... unit_cost_usd  unit_price_usd  subcategory_key  \
0          Contoso  ...         31.27           68.00              406   
1  Adventure Works  ...        166.20          326.00              301   
2  Adventure Works  ...        527.53         1592.20              201   
3  Adventure Works  ...        527.53         1592.20              201   
4  Adventure Works  ...        215.68          469.00              303   
5  Adventure Works  ...        215.68          469.00              303   
6  Adventure Works  ...        215.68          469.00              303   
7          Contoso  ...         26.67           57.99              602   
8          Contoso  ...         26.67           57.99              602   
9          Contoso  ...          2.07            4.06              505   

                        subcategory category_key  \
0  Cameras & Camcorders Accessories            4   
1                           Laptops            3   
2                       Televisions            2   
3                       Televisions            2   
4                          Desktops            3   
5                          Desktops            3   
6                          Desktops            3   
7                         Movie DVD            6   
8                         Movie DVD            6   
9           Cell phones Accessories            5   

                        category country    state square_meters  open_date  
0         Cameras and camcorders  Canada  Nunavut        1210.0   4/4/2015  
1                      Computers  Canada  Nunavut        1210.0   4/4/2015  
2                   TV and Video  Canada  Nunavut        1210.0   4/4/2015  
3                   TV and Video  Canada  Nunavut        1210.0   4/4/2015  
4                      Computers  Canada  Nunavut        1210.0   4/4/2015  
5                      Computers  Canada  Nunavut        1210.0   4/4/2015  
6                      Computers  Canada  Nunavut        1210.0   4/4/2015  
7  Music, Movies and Audio Books  Canada  Nunavut        1210.0   4/4/2015  
8  Music, Movies and Audio Books  Canada  Nunavut        1210.0   4/4/2015  
9                    Cell phones  Canada  Nunavut        1210.0   4/4/2015  

[10 r

In [64]:
df.isna().sum()

order_number           0
line_item              0
order_date             0
customer_key           0
store_key              0
product_key            0
quantity               0
currency_code          0
product_name           0
brand                  0
color                  0
unit_cost_usd          0
unit_price_usd         0
subcategory_key        0
subcategory            0
category_key           0
category               0
country                0
state                  0
square_meters      13165
open_date              0
dtype: int64

In [65]:
online_stores = df.loc[df.store_key == 0]
online_stores.shape

(13165, 21)

As we saw before, we have an online distribution channel that does not have a physical location. For that reason, we have the same number of missing values in our dataset for square meterse as we the number of online transactions. 

13,165 values are missing from square meteres because they were purchases made online and not in a store. We will leave these values the way they are and address this during the modeling stage. 

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 62884 entries, 0 to 62883
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   order_number     62884 non-null  int64  
 1   line_item        62884 non-null  int64  
 2   order_date       62884 non-null  object 
 3   customer_key     62884 non-null  int64  
 4   store_key        62884 non-null  int64  
 5   product_key      62884 non-null  int64  
 6   quantity         62884 non-null  int64  
 7   currency_code    62884 non-null  object 
 8   product_name     62884 non-null  object 
 9   brand            62884 non-null  object 
 10  color            62884 non-null  object 
 11  unit_cost_usd    62884 non-null  float64
 12  unit_price_usd   62884 non-null  float64
 13  subcategory_key  62884 non-null  int64  
 14  subcategory      62884 non-null  object 
 15  category_key     62884 non-null  int64  
 16  category         62884 non-null  object 
 17  country     

**Preliminary Feature Engineering**

While we will dive deeper into engineering other features such as intercation and polynomial features, there are some obvious ones we can make now, including building our target feature. 

First we need to cirecle back and remember that we have multiple currencies, however, our categories unit_cost and unit_price have already been converted into USD, so we will continue using USD as the baseline for analysis. 

- Revenue
- Costs
- Profit Margin 

In [67]:
df['revenue'] = df['unit_price_usd'] * df['quantity']

In [68]:
df['costs'] = df['unit_cost_usd'] * df['quantity']

In [69]:
df['profit_margin'] = df['revenue'] - df['costs']

In [70]:
df.head()

order_number  line_item  order_date  customer_key  store_key  product_key  \
0        366000          1    1/1/2016        265598         10         1304   
1       1896020          1   3/10/2020        302645         10          373   
2       1298021          1   7/21/2018        363584         10          163   
3       1506050          4   2/14/2019        391900         10          163   
4       1452075          3  12/22/2018        211746         10          421   

   quantity currency_code                                product_name  \
0         1           CAD             Contoso Lens Adapter M450 White   
1         1           CAD       Adventure Works Laptop8.9 E0890 White   
2         8           CAD    Adventure Works 52" LCD HDTV X790W White   
3         4           CAD    Adventure Works 52" LCD HDTV X790W White   
4         2           CAD  Adventure Works Desktop PC3.0 MS300 Silver   

             brand  ...                       subcategory  category_key  \
0          Contoso  ...  Cameras & Camcorders Accessories             4   
1  Adventure Works  ...                           Laptops             3   
2  Adventure Works  ...                       Televisions             2   
3  Adventure Works  ...                       Televisions             2   
4  Adventure Works  ...                          Desktops             3   

                 category  country    state  square_meters open_date  revenue  \
0  Cameras and camcorders   Canada  Nunavut         1210.0  4/4/2015     68.0   
1               Computers   Canada  Nunavut         1210.0  4/4/2015    326.0   
2            TV and Video   Canada  Nunavut         1210.0  4/4/2015  12737.6   
3            TV and Video   Canada  Nunavut         1210.0  4/4/2015   6368.8   
4               Computers   Canada  Nunavut         1210.0  4/4/2015    938.0   

     costs  profit_margin  
0    31.27          36.73  
1   166.20         159.80  
2  4220.24        8517.36  
3  2110.12        4258.68  
4   431.36         506.64  

[5 rows x 24 columns]

Let's go back to the unique customer we looked at earlier to see if the math adds up:

In [71]:
unique_customer = df.loc[df.customer_key == 265598.0]
unique_customer[['order_number', 'customer_key', 'store_key', 'unit_cost_usd', 'unit_price_usd', 'quantity', 'revenue', 'costs', 'profit_margin']]

order_number  customer_key  store_key  unit_cost_usd  unit_price_usd  \
0            366000        265598         10          31.27           68.00   
8716        1749017        265598          0          27.59           59.99   
14128       1749017        265598          0          91.95          199.95   
16196       1749017        265598          0          23.86           46.80   
36465        891000        265598          9          43.04          129.90   
36610        891000        265598          9         166.20          326.00   
36711        891000        265598          9          45.98           99.99   
37236        891000        265598          9          72.66          158.00   

       quantity  revenue   costs  profit_margin  
0             1    68.00   31.27          36.73  
8716          1    59.99   27.59          32.40  
14128         1   199.95   91.95         108.00  
16196         2    93.60   47.72          45.88  
36465         1   129.90   43.04          86.86  
36610         2   652.00  332.40         319.60  
36711         1    99.99   45.98          54.01  
37236         1   158.00   72.66          85.34

The math is mathing

Lastly, we will import the customer dataset to include features such as gender, birthdate (to get age), in order to see if these features have strong influence on our target **profit margin** 

# Third Merge

In [72]:
file_path = '~/Desktop/Customers.csv'

customers_df = pd.read_csv(file_path, encoding='ISO-8859-1')

In [73]:
customers_df.head()

customer_key  gender               name            city state_code  \
0           301  Female      Lilly Harding  WANDEARAH EAST         SA   
1           325  Female       Madison Hull      MOUNT BUDD         WA   
2           554  Female      Claire Ferres       WINJALLOK        VIC   
3           786    Male  Jai Poltpalingada    MIDDLE RIVER         SA   
4          1042    Male    Aidan Pankhurst   TAWONGA SOUTH        VIC   

               state zip_code    country  continent    birthday  
0    South Australia     5523  Australia  Australia    7/3/1939  
1  Western Australia     6522  Australia  Australia   9/27/1979  
2           Victoria     3380  Australia  Australia   5/26/1947  
3    South Australia     5223  Australia  Australia   9/17/1957  
4           Victoria     3698  Australia  Australia  11/19/1965

In [74]:
customers_df.isna().sum()

customer_key     0
gender           0
name             0
city             0
state_code      10
state            0
zip_code         0
country          0
continent        0
birthday         0
dtype: int64

In [75]:
extract_columns = ['customer_key', 'city', 'gender', 'zip_code', 'birthday']

customers_df = customers_df[extract_columns]

In [76]:
customers_df.columns

Index(['customer_key', 'city', 'gender', 'zip_code', 'birthday'], dtype='object')

We won't be needing state_code anyway

In [77]:
df = pd.merge(df, customers_df, on='customer_key', how='inner')

In [78]:
df.head()

order_number  line_item  order_date  customer_key  store_key  product_key  \
0        366000          1    1/1/2016        265598         10         1304   
1       1749017          1  10/15/2019        265598          0         1619   
2       1749017          3  10/15/2019        265598          0           52   
3       1749017          2  10/15/2019        265598          0         1764   
4        891000          2    6/9/2017        265598          9          174   

   quantity currency_code                                  product_name  \
0         1           CAD               Contoso Lens Adapter M450 White   
1         1           CAD                  Contoso DVD Player M130 Grey   
2         1           CAD           WWI 2GB Pulse Smart pen M100 Silver   
3         2           CAD  MGS Age of Empires II: The Age of Kings E126   
4         1           CAD                          SV 22xDVD M600 Black   

                  brand  ...                  state  square_meters  open_date  \
0               Contoso  ...                Nunavut         1210.0   4/4/2015   
1               Contoso  ...                 Online            NaN   1/1/2010   
2  Wide World Importers  ...                 Online            NaN   1/1/2010   
3         Tailspin Toys  ...                 Online            NaN   1/1/2010   
4      Southridge Video  ...  Northwest Territories         1500.0   3/4/2005   

   revenue  costs  profit_margin    city gender zip_code   birthday  
0    68.00  31.27          36.73  London   Male  N5W 5K6  3/23/1971  
1    59.99  27.59          32.40  London   Male  N5W 5K6  3/23/1971  
2   199.95  91.95         108.00  London   Male  N5W 5K6  3/23/1971  
3    93.60  47.72          45.88  London   Male  N5W 5K6  3/23/1971  
4   129.90  43.04          86.86  London   Male  N5W 5K6  3/23/1971  

[5 rows x 28 columns]

In [79]:
df.isna().sum()

order_number           0
line_item              0
order_date             0
customer_key           0
store_key              0
product_key            0
quantity               0
currency_code          0
product_name           0
brand                  0
color                  0
unit_cost_usd          0
unit_price_usd         0
subcategory_key        0
subcategory            0
category_key           0
category               0
country                0
state                  0
square_meters      13165
open_date              0
revenue                0
costs                  0
profit_margin          0
city                   0
gender                 0
zip_code               0
birthday               0
dtype: int64

Let's take a look at our customer age groups and create necessary columns:

In [80]:
#Take a look at our date Range 

print(df.order_date.min())
print(df.order_date.max())

1/1/2016
9/9/2020


In [81]:
print(df.order_date.dtype)

object


In [82]:
#Convert order date to datetime

df.order_date = pd.to_datetime(df['order_date'])
print(df['order_date'].dtype)

datetime64[ns]


In [83]:
#Convert birthday to datetime

df['birthday'] = pd.to_datetime(df['birthday'])
print(df['birthday'].dtype)

datetime64[ns]


In [84]:
#Create birth_year column
df['birth_year'] = df['birthday'].dt.year

df['birth_year'].unique()

array([1971, 1985, 1977, 1994, 1947, 1957, 1958, 1950, 1987, 1952, 1980,
       1949, 1998, 1975, 1993, 1960, 1938, 1979, 1935, 1948, 1946, 1942,
       1967, 1953, 1951, 1968, 1944, 1959, 2000, 1965, 1978, 1982, 2001,
       1963, 1966, 1973, 1941, 1988, 1986, 1983, 1939, 1974, 1991, 1961,
       1964, 1954, 1995, 1999, 1984, 1943, 1981, 1937, 1972, 1976, 1997,
       1989, 1970, 2002, 1969, 1936, 1955, 1945, 1990, 1996, 1992, 1962,
       1956, 1940])

In [85]:
print(df['birth_year'].min())
print(df['birth_year'].max())

1935
2002


In [86]:
df['order_date'].dt.year.unique()

array([2016, 2019, 2017, 2020, 2018, 2021])

How do we have an order year of 2021?

In [87]:
# Filter for orders made in the year 2021
filtered_df = df.loc[df['order_date'].dt.year == 2021]

# Display the filtered DataFrame
print(filtered_df)

       order_number  line_item order_date  customer_key  store_key  \
143         2199014          1 2021-01-07        242542          0   
223         2241011          6 2021-02-18        224111         10   
224         2241011          2 2021-02-18        224111         10   
226         2241012          2 2021-02-18        224111         10   
232         2241011          3 2021-02-18        224111         10   
...             ...        ...        ...           ...        ...   
62423       2233004          2 2021-02-10         45814          1   
62424       2233004          3 2021-02-10         45814          1   
62795       2211006          2 2021-01-19        801780         34   
62796       2211006          1 2021-01-19        801780         34   
62881       2233009          1 2021-02-10        416365         26   

       product_key  quantity currency_code  \
143            446         3           CAD   
223           1637         3           CAD   
224           1615   

In [88]:
#Create order_year column and check the years
df['order_year'] = df['order_date'].dt.year

df['order_year'].unique()

array([2016, 2019, 2017, 2020, 2018, 2021])

In [89]:
#Create order_month column and check the months
df['order_month'] = df['order_date'].dt.month

df['order_month'].isna().sum()

0

In [90]:
print(df['order_year'].shape)
print(df['order_month'].shape)

(62884,)
(62884,)


In [91]:
#Create age_purcahse column to include just the age in years of the customer at time of purchase
df['age_purchase'] = df['order_year'] - df['birth_year']

df['age_purchase'].unique()

array([45, 48, 46, 35, 34, 41, 42, 43, 25, 24, 71, 70, 69, 62, 61, 58, 66,
       30, 67, 40, 19, 21, 23, 22, 27, 39, 57, 60, 78, 82, 80, 18, 20, 68,
       64, 83, 84, 81, 72, 73, 37, 74, 49, 44, 65, 51, 63, 52, 54, 55, 38,
       36, 56, 16, 59, 76, 32, 33, 53, 28, 26, 47, 17, 77, 29, 31, 50, 75,
       79, 85, 15, 86, 14])

In [92]:
print(df['age_purchase'].min())
print(df['age_purchase'].max())

14
86


The customer age range is wide, we will bin them to make analyzing demographis easier:

Bins:
- 14-25
- 26-35
- 36-45
- 46-55
- 56-65
- 66-75
- 76-86

In [93]:
#Create the edges of your age groups
bins = [13, 25, 35, 45, 55, 65, 75, 86] 

#Create the label names
labels = ['14-25', '26-35', '36-45', '46-55', '56-65', '66-75', '76-86']

#Create the age group column
df['age_group'] = pd.cut(df['age_purchase'], bins=bins, labels=labels, right=True)

print(df['age_group'].unique())

['36-45', '46-55', '26-35', '14-25', '66-75', '56-65', '76-86']
Categories (7, object): ['14-25' < '26-35' < '36-45' < '46-55' < '56-65' < '66-75' < '76-86']


In [94]:
df['age_group'].value_counts()

36-45    9437
66-75    9428
46-55    9373
56-65    9326
26-35    9271
14-25    8562
76-86    7487
Name: age_group, dtype: int64

In [95]:
df['age_group'].isna().sum()

0

**Note -** We've made them age_group column into bins that contain the actual age group each individual person falls in as an object data type. This is so that we can use value counts easier during our exploratory data anlysis. 
We will likely encode the age column in the pre-processing stage in order to prepare it for our ML models. 

In [96]:
print(df['open_date'].dtype)

object


In [97]:
df['open_date'] = pd.to_datetime(df['open_date'])
print(df['open_date'].dtype)

datetime64[ns]


In [98]:
df['open_date'].nunique()

23

In [99]:
df.head(1)

order_number  line_item order_date  customer_key  store_key  product_key  \
0        366000          1 2016-01-01        265598         10         1304   

   quantity currency_code                     product_name    brand  ...  \
0         1           CAD  Contoso Lens Adapter M450 White  Contoso  ...   

  profit_margin    city  gender  zip_code   birthday  birth_year order_year  \
0         36.73  London    Male   N5W 5K6 1971-03-23        1971       2016   

  order_month age_purchase  age_group  
0           1           45      36-45  

[1 rows x 33 columns]

In [100]:
df.rename(columns={'revenue': 'sales', 'profit_margin': 'profit'}, inplace=True)

df.columns

Index(['order_number', 'line_item', 'order_date', 'customer_key', 'store_key',
       'product_key', 'quantity', 'currency_code', 'product_name', 'brand',
       'color', 'unit_cost_usd', 'unit_price_usd', 'subcategory_key',
       'subcategory', 'category_key', 'category', 'country', 'state',
       'square_meters', 'open_date', 'sales', 'costs', 'profit', 'city',
       'gender', 'zip_code', 'birthday', 'birth_year', 'order_year',
       'order_month', 'age_purchase', 'age_group'],
      dtype='object')

**Years Open Column**

In [101]:
df['order_date'].max()

Timestamp('2021-02-20 00:00:00')

In [102]:
df['years_open'] = df['order_date'] - df['open_date']

df['years_open'].head()

0    272 days
1   3574 days
2   3574 days
3   3574 days
4   4480 days
Name: years_open, dtype: timedelta64[ns]

In [103]:
df['years_open'] = (df['order_date'] - df['open_date']).dt.days / 365
df['years_open'].head()

0     0.745205
1     9.791781
2     9.791781
3     9.791781
4    12.273973
Name: years_open, dtype: float64

In [104]:
df.to_csv('~/Desktop/retailer_data.csv', index=False)